message/infrastructure/sql/subscriber.go

```diff
@@ -0,0 +1,326 @@
+	MessagesTable string
+	// MessageOffsetsTable is the name of the table that stores the offsets of messages read by each consumer group.
+	// Defaults to `offsets_acked`.
+	MessageOffsetsTable string
+
+	// SchemaAdapter provides the schema-dependent queries and arguments for them, based on topic/message etc.
+	SchemaAdapter SchemaAdapter
+}
+
+func (c *SubscriberConfig) setDefaults() {
+	if c.Logger == nil {
+		c.Logger = watermill.NopLogger{}
+	}
+	if c.PollInterval == 0 {
+		c.PollInterval = time.Second
+	}
+	if c.ResendInterval == 0 {
+		c.ResendInterval = time.Second
+	}
+	if c.MessagesTable == "" {
+		c.MessagesTable = "messages"
+	}
+	if c.MessageOffsetsTable == "" {
+		c.MessageOffsetsTable = "offsets_acked"
+	}
+}
+
+func (c SubscriberConfig) validate() error {
+	if c.PollInterval <= 0 {
+		return errors.New("poll interval must be a positive duration")
+	}
+	if c.ResendInterval <= 0 {
+		return errors.New("resend interval must be a positive duration")
+	}
+	if c.SchemaAdapter == nil {
+		return errors.New("schema adapter is nil")
+	}
+
+	return nil
+}
+
+// Subscriber makes SELECT queries on the chosen table with the interval defined in the config.
+// The rows are unmarshaled into Watermill messages.
+type Subscriber struct {
+	db     *sql.DB
+	config SubscriberConfig
+
+	subscribeWg *sync.WaitGroup
+	closing     chan struct{}
+	closed      bool
+}
+
+func NewSubscriber(db *sql.DB, conf SubscriberConfig) (*Subscriber, error) {
+	if db == nil {
+		return nil, errors.New("db is nil")
+	}
+	conf.setDefaults()
+	err := conf.validate()
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid config")
+	}
+
+	sub := &Subscriber{
+		db:     db,
+		config: conf,
+
+		subscribeWg: &sync.WaitGroup{},
+		closing:     make(chan struct{}),
+	}
+
+	return sub, nil
+}
+
+func (s *Subscriber) Subscribe(ctx context.Context, topic string) (o <-chan *message.Message, err error) {
+	if s.closed {
+		return nil, ErrSubscriberClosed
+	}
+
+	if err = sanitizeTopicName(topic); err != nil {
+		return nil, err
+	}
+
+	// the information about closing the subscriber is propagated through ctx
+	ctx, cancel := context.WithCancel(ctx)
+	out := make(chan *message.Message)
+
+	s.subscribeWg.Add(1)
+	go func() {
+		s.consume(ctx, topic, out)
+		close(out)
+		cancel()
+	}()
+
+	return out, nil
+}
+
+func (s *Subscriber) consume(ctx context.Context, topic string, out chan *message.Message) {
+	defer s.subscribeWg.Done()
+
+	logger := s.config.Logger.With(watermill.LogFields{
+		"topic":          topic,
+		"consumer_group": s.config.ConsumerGroup,
+	})
+
+	for {
+		select {
+		case <-s.closing:
+			logger.Info("Discarding queued message, subscriber closing", nil)
+			return
+
+		case <-ctx.Done():
+			logger.Info("Stopping consume, context canceled", nil)
+			return
+
+		default:
+			// go on querying
+		}
+
+		err := s.query(ctx, topic, out, logger)
+		if err != nil {
+			logger.Error("Error querying for message", err, nil)
```

Comment:
maybe some configureable sleep, to not kill db? :P

Reply:
RetryInterval de25f1885ee23ee91c57f23b16957eab61fa5900

Reply:
actually used in 14532ac9e8f55a650f47ecfeb5e4824681748c29